# Wine Quality Analysis & Feature Engineering

This Project analyzes chemical properties of red wine to understand the key factors influencing wine quality.


## Business Problem

Wine producers aim to understand which chemical properties most strongly influence perceived wine quality.

Objective:
- Identify key drivers of quality
- Engineer meaningful features
- Extract actionable insights


| Column Name              | Description                                                               | Why It Matters                                                                |
| ------------------------ | ------------------------------------------------------------------------- | ----------------------------------------------------------------------------- |
| **fixed acidity**        | Concentration of non-volatile acids (e.g., tartaric acid) in the wine.    | Contributes to overall taste structure and stability.                         |
| **volatile acidity**     | Amount of volatile acids (mainly acetic acid) that can evaporate.         | High levels may produce an undesirable vinegar-like taste and reduce quality. |
| **citric acid**          | Quantity of citric acid present in the wine.                              | Adds freshness and enhances flavor balance.                                   |
| **residual sugar**       | Remaining sugar after fermentation is complete.                           | Influences sweetness level and wine style (dry vs sweet).                     |
| **chlorides**            | Amount of salt (sodium chloride) in the wine.                             | Excessive levels may negatively affect taste.                                 |
| **free sulfur dioxide**  | Free form of sulfur dioxide that prevents oxidation and microbial growth. | Essential for preservation but harmful if excessive.                          |
| **total sulfur dioxide** | Total amount of sulfur dioxide (free + bound).                            | Indicates overall preservation level and stability.                           |
| **density**              | Density of the wine (related to sugar and alcohol content).               | Higher alcohol lowers density; useful for detecting fermentation effects.     |
| **pH**                   | Acidity level of the wine (lower = more acidic).                          | Affects taste, color stability, and microbial activity.                       |
| **sulphates**            | Potassium sulphate concentration in wine.                                 | Can enhance fermentation and potentially improve quality.                     |
| **alcohol**              | Alcohol percentage in the wine.                                           | Often positively associated with higher wine quality.                         |
| **quality**              | Sensory quality score (typically between 0–10).                           | Target variable representing expert evaluation of wine quality.               |
| **Id**                   | Unique identifier for each wine sample.                                   | Not analytically useful; typically removed during preprocessing.              |


In [1]:
# import Libraries
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import plotly.express as px


In [ ]:

df = pd.read_csv("D:\\Microsoft-Machine-Learning-Engineer-DEPI-Round4\\Data\\WineQT.csv") 
df

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.4             0.700         0.00             1.9      0.076   
1               7.8             0.880         0.00             2.6      0.098   
2               7.8             0.760         0.04             2.3      0.092   
3              11.2             0.280         0.56             1.9      0.075   
4               7.4             0.700         0.00             1.9      0.076   
...             ...               ...          ...             ...        ...   
1138            6.3             0.510         0.13             2.3      0.076   
1139            6.8             0.620         0.08             1.9      0.068   
1140            6.2             0.600         0.08             2.0      0.090   
1141            5.9             0.550         0.10             2.2      0.062   
1142            5.9             0.645         0.12             2.0      0.075   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    11.0                  34.0  0.99780  3.51       0.56   
1                    25.0                  67.0  0.99680  3.20       0.68   
2                    15.0                  54.0  0.99700  3.26       0.65   
3                    17.0                  60.0  0.99800  3.16       0.58   
4                    11.0                  34.0  0.99780  3.51       0.56   
...                   ...                   ...      ...   ...        ...   
1138                 29.0                  40.0  0.99574  3.42       0.75   
1139                 28.0                  38.0  0.99651  3.42       0.82   
1140                 32.0                  44.0  0.99490  3.45       0.58   
1141                 39.0                  51.0  0.99512  3.52       0.76   
1142                 32.0                  44.0  0.99547  3.57       0.71   

      alcohol  quality    Id  
0         9.4        5     0  
1         9.8        5     1  
2         9.8        5     2  
3         9.8        6     3  
4         9.4        5     4  
...       ...      ...   ...  
1138     11.0        6  1592  
1139      9.5        6  1593  
1140     10.5        5  1594  
1141     11.2        6  1595  
1142     10.2        5  1597  

[1143 rows x 13 columns]

## Data Understanding

In [3]:
df.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  Id  
0      9.4        5   0  
1      9.8        5   1  
2      9.8        5   2  
3      9.8        6   3  
4      9.4        5   4

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1143 entries, 0 to 1142
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         1143 non-null   float64
 1   volatile acidity      1143 non-null   float64
 2   citric acid           1143 non-null   float64
 3   residual sugar        1143 non-null   float64
 4   chlorides             1143 non-null   float64
 5   free sulfur dioxide   1143 non-null   float64
 6   total sulfur dioxide  1143 non-null   float64
 7   density               1143 non-null   float64
 8   pH                    1143 non-null   float64
 9   sulphates             1143 non-null   float64
 10  alcohol               1143 non-null   float64
 11  quality               1143 non-null   int64  
 12  Id                    1143 non-null   int64  
dtypes: float64(11), int64(2)
memory usage: 116.2 KB


In [8]:
df.shape

(1143, 13)

In [9]:
df.describe()

fixed acidity  volatile acidity  citric acid  residual sugar  \
count    1143.000000       1143.000000  1143.000000     1143.000000   
mean        8.311111          0.531339     0.268364        2.532152   
std         1.747595          0.179633     0.196686        1.355917   
min         4.600000          0.120000     0.000000        0.900000   
25%         7.100000          0.392500     0.090000        1.900000   
50%         7.900000          0.520000     0.250000        2.200000   
75%         9.100000          0.640000     0.420000        2.600000   
max        15.900000          1.580000     1.000000       15.500000   

         chlorides  free sulfur dioxide  total sulfur dioxide      density  \
count  1143.000000          1143.000000           1143.000000  1143.000000   
mean      0.086933            15.615486             45.914698     0.996730   
std       0.047267            10.250486             32.782130     0.001925   
min       0.012000             1.000000              6.000000     0.990070   
25%       0.070000             7.000000             21.000000     0.995570   
50%       0.079000            13.000000             37.000000     0.996680   
75%       0.090000            21.000000             61.000000     0.997845   
max       0.611000            68.000000            289.000000     1.003690   

                pH    sulphates      alcohol      quality           Id  
count  1143.000000  1143.000000  1143.000000  1143.000000  1143.000000  
mean      3.311015     0.657708    10.442111     5.657043   804.969379  
std       0.156664     0.170399     1.082196     0.805824   463.997116  
min       2.740000     0.330000     8.400000     3.000000     0.000000  
25%       3.205000     0.550000     9.500000     5.000000   411.000000  
50%       3.310000     0.620000    10.200000     6.000000   794.000000  
75%       3.400000     0.730000    11.100000     6.000000  1209.500000  
max       4.010000     2.000000    14.900000     8.000000  1597.000000

### Data Cleaning

In [10]:
df.drop(columns=['Id'], inplace=True)
df.duplicated().sum()


np.int64(125)

In [11]:
df = df.drop_duplicates()

In [12]:

df.duplicated().sum()


np.int64(0)

In [13]:
df.isnull().sum()

fixed acidity           0
volatile acidity        0
citric acid             0
residual sugar          0
chlorides               0
free sulfur dioxide     0
total sulfur dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
dtype: int64

### Feature Engineering

In [15]:
# Create a new column 'quality_label' based on the 'quality' column
df["quality_label"] = pd.cut(
    df["quality"],
    bins=[0,5,6,10],
    labels=["Low","Medium","High"]
)
df.head()

C:\Users\Basmala\AppData\Local\Temp\ipykernel_14508\1989343775.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["quality_label"] = pd.cut(


fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
5            7.4              0.66         0.00             1.8      0.075   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
5                 13.0                  40.0   0.9978  3.51       0.56   

   alcohol  quality quality_label  
0      9.4        5           Low  
1      9.8        5           Low  
2      9.8        5           Low  
3      9.8        6        Medium  
5      9.4        5           Low

In [16]:
# Total Acidity Feature
df["total_acidity"] = (
    df["fixed acidity"] +
    df["volatile acidity"] +
    df["citric acid"]
)
df.head()

C:\Users\Basmala\AppData\Local\Temp\ipykernel_14508\1060423828.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["total_acidity"] = (


fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
5            7.4              0.66         0.00             1.8      0.075   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
5                 13.0                  40.0   0.9978  3.51       0.56   

   alcohol  quality quality_label  total_acidity  
0      9.4        5           Low           8.10  
1      9.8        5           Low           8.68  
2      9.8        5           Low           8.60  
3      9.8        6        Medium          12.04  
5      9.4        5           Low           8.06

In [17]:
# Acidity Balance Ratio
df["acidity_balance"] = df["fixed acidity"] / (df["volatile acidity"] + 1e-5)
df.head()

C:\Users\Basmala\AppData\Local\Temp\ipykernel_14508\1026632161.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["acidity_balance"] = df["fixed acidity"] / (df["volatile acidity"] + 1e-5)


fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
5            7.4              0.66         0.00             1.8      0.075   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
5                 13.0                  40.0   0.9978  3.51       0.56   

   alcohol  quality quality_label  total_acidity  acidity_balance  
0      9.4        5           Low           8.10        10.571278  
1      9.8        5           Low           8.68         8.863536  
2      9.8        5           Low           8.60        10.263023  
3      9.8        6        Medium          12.04        39.998571  
5      9.4        5           Low           8.06        11.211951

In [18]:
# Sulfur Efficiency
df["sulfur_efficiency"] = (
    df["free sulfur dioxide"] /
    (df["total sulfur dioxide"] + 1e-5)
)
df.head()

C:\Users\Basmala\AppData\Local\Temp\ipykernel_14508\468677309.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["sulfur_efficiency"] = (


fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
5            7.4              0.66         0.00             1.8      0.075   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
5                 13.0                  40.0   0.9978  3.51       0.56   

   alcohol  quality quality_label  total_acidity  acidity_balance  \
0      9.4        5           Low           8.10        10.571278   
1      9.8        5           Low           8.68         8.863536   
2      9.8        5           Low           8.60        10.263023   
3      9.8        6        Medium          12.04        39.998571   
5      9.4        5           Low           8.06        11.211951   

   sulfur_efficiency  
0           0.323529  
1           0.373134  
2           0.277778  
3           0.283333  
5           0.325000

In [19]:
# Fermentation Strength
df["fermentation_strength"] = df["alcohol"] / df["density"]
df.head()

C:\Users\Basmala\AppData\Local\Temp\ipykernel_14508\31881297.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["fermentation_strength"] = df["alcohol"] / df["density"]


fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
5            7.4              0.66         0.00             1.8      0.075   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
5                 13.0                  40.0   0.9978  3.51       0.56   

   alcohol  quality quality_label  total_acidity  acidity_balance  \
0      9.4        5           Low           8.10        10.571278   
1      9.8        5           Low           8.68         8.863536   
2      9.8        5           Low           8.60        10.263023   
3      9.8        6        Medium          12.04        39.998571   
5      9.4        5           Low           8.06        11.211951   

   sulfur_efficiency  fermentation_strength  
0           0.323529               9.420726  
1           0.373134               9.831461  
2           0.277778               9.829488  
3           0.283333               9.819639  
5           0.325000               9.420726

In [20]:
# Sugar-Alcohol Interaction
df["sugar_alcohol_ratio"] = (
    df["residual sugar"] / (df["alcohol"] + 1e-5)
)
df.head()

C:\Users\Basmala\AppData\Local\Temp\ipykernel_14508\2203600518.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["sugar_alcohol_ratio"] = (


fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
5            7.4              0.66         0.00             1.8      0.075   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
5                 13.0                  40.0   0.9978  3.51       0.56   

   alcohol  quality quality_label  total_acidity  acidity_balance  \
0      9.4        5           Low           8.10        10.571278   
1      9.8        5           Low           8.68         8.863536   
2      9.8        5           Low           8.60        10.263023   
3      9.8        6        Medium          12.04        39.998571   
5      9.4        5           Low           8.06        11.211951   

   sulfur_efficiency  fermentation_strength  sugar_alcohol_ratio  
0           0.323529               9.420726             0.202127  
1           0.373134               9.831461             0.265306  
2           0.277778               9.829488             0.234694  
3           0.283333               9.819639             0.193877  
5           0.325000               9.420726             0.191489

In [21]:
# Sweetness Category
df["sweetness"] = pd.cut(
    df["residual sugar"],
    bins=[0,2,5,20],
    labels=["Dry","Semi-Sweet","Sweet"]
)
df.head()

C:\Users\Basmala\AppData\Local\Temp\ipykernel_14508\124896736.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["sweetness"] = pd.cut(


fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
5            7.4              0.66         0.00             1.8      0.075   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
5                 13.0                  40.0   0.9978  3.51       0.56   

   alcohol  quality quality_label  total_acidity  acidity_balance  \
0      9.4        5           Low           8.10        10.571278   
1      9.8        5           Low           8.68         8.863536   
2      9.8        5           Low           8.60        10.263023   
3      9.8        6        Medium          12.04        39.998571   
5      9.4        5           Low           8.06        11.211951   

   sulfur_efficiency  fermentation_strength  sugar_alcohol_ratio   sweetness  
0           0.323529               9.420726             0.202127         Dry  
1           0.373134               9.831461             0.265306  Semi-Sweet  
2           0.277778               9.829488             0.234694  Semi-Sweet  
3           0.283333               9.819639             0.193877         Dry  
5           0.325000               9.420726             0.191489         Dry

### Exploratory Data Analysis (EDA)

In [ ]:
# Quality Distribution

px.histogram(
    df,
    x="quality",
    nbins=10,
    title="Distribution of Wine Quality",
    text_auto=True
)


Insight:

- Most wines are rated between 5 and 6.

- High-quality wines (7+) are underrepresented.

- The dataset is slightly imbalanced toward medium quality wines.

In [24]:
# Alcohol vs Quality
px.box(
    df,
    x="quality",
    y="alcohol",
    title="Alcohol Content by Quality Score"
)

Insight:

- Higher quality wines tend to have higher alcohol levels.

- Alcohol shows a clear positive relationship with quality.

In [25]:
# Volatile Acidity vs Quality
px.box(
    df,
    x="quality",
    y="volatile acidity",
    title="Volatile Acidity by Quality Score"
)

Insight:

- Higher volatile acidity is associated with lower quality.

- Excessive acetic acid negatively impacts wine perception.

In [26]:
# Sulphates vs Quality
px.box(
    df,
    x="quality",
    y="sulphates",
    title="Sulphates by Quality Score"
)

Insight:

- Moderate sulphate levels appear more frequently in higher-quality wines.

- Sulphates may contribute to fermentation stability and quality improvement.

In [ ]:
# Correlation Matrix
corr = df.corr(numeric_only=True)
px.imshow(
    corr,
    text_auto=True,
    title="Correlation Matrix"
    
)



Insight:

- Alcohol positively correlates with quality.

- Volatile acidity negatively correlates with quality.

- Density inversely correlates with alcohol.

In [29]:
# Acidity Ratio Distribution
px.histogram(
    df,
    x="acidity_balance",
    nbins=50,
    title="Distribution of Acidity Balance Ratio"
)


In [33]:
# Sweetness vs Quality Category
px.box(
    df,
    x="sweetness",
    y="quality",
    title="Sweetness Level vs Quality Score"
)



Insight:

- Most high-quality wines are Sweet.

- Sweetness is  a strong quality determinant.

## Key Insights
- Most wines are rated as medium quality (5–6).
- Alcohol is the strongest positive driver of quality.
- Volatile acidity negatively impacts wine ratings.
- Density decreases as alcohol increases.
- Engineered chemical balance ratios provide clearer insights.
- Sweetness has unlimited influence on perceived quality.
- Quality appears influenced more by chemical balance less than sugar concentration.


## Executive Summary

This project analyzed the chemical properties of red wine to identify key drivers of wine quality.

The dataset includes 1,143 samples with 11 chemical attributes and a quality rating score.

Key Findings:

- Alcohol content is the strongest positive predictor of wine quality.
- Volatile acidity negatively impacts sensory ratings.
- Most wines fall into medium quality levels (5–6).
- Feature engineering revealed that chemical balance ratios provide deeper interpretability than standalone features.

These insights can support winemakers in improving fermentation control, product differentiation, and premium positioning strategies.
